include/ofi_recvwin.h

```diff
@@ -123,8 +123,14 @@ ofi_recvwin_slide(struct name *recvq)					\
 #define ofi_recvwin_exp_inc(rq)		((rq)->exp_msg_id++)
 #define ofi_recvwin_is_exp(rq, id)	((rq)->exp_msg_id == id)
 #define ofi_recvwin_next_exp_id(rq)	((rq)->exp_msg_id)
-#define ofi_recvwin_is_delayed(rq, id)	((rq)->exp_msg_id > id)
-#define ofi_recvwin_is_allowed(rq, id)	(id >= rq->exp_msg_id \
-					&& id < (rq->win_size + rq->exp_msg_id))
+/*
+ * Check if the reorder ID range is split if exp_msg_id is about to wraparound.
+ */
+#define ofi_recvwin_id_split(rq)	((rq)->exp_msg_id + (rq)->win_size < (rq)->win_size)
+#define ofi_recvwin_is_allowed(rq, id)	((id >= (rq)->exp_msg_id && \
+					  id < (rq)->win_size + (rq)->exp_msg_id) || \
+					 (ofi_recvwin_id_split(rq) && \
+					  (id >= (rq)->exp_msg_id || \
+					  (id < ((rq)->exp_msg_id + (rq)->win_size)))))
```

Comment:
I'm hoping that we can find a simpler or faster way of performing this check.  Can something like the following work (adds some generic macros first):
```
#define ofi_val_gt(x, y) \
    ((int64_t) ((uint64_t) (x) - (uint64_t) (y))) > 0)
#define ofi_val_ge(x, y) \
    ((int64_t) ((uint64_t) (x) - (uint64_t) (y))) >= 0)
#define ofi_val_lt(x, y) ofi_val_gt(y, x)

#define ofi_val_inrange(start, length, value)
    ofi_val_ge(value, start) && ofi_val_lt(value, start + length)
```

Reply:
Sure, that will work. We'll have to use `int32_t` / `uint32_t` for EFA though. Should I just define two sets of macros, one for 32 bit and one for 64 bit? I'm not sure if there's a way to write the macro to work for either 32 or 64 bit values.

Reply:
We can define 2 sets of macros if needed (val32 / val64).  I was hoping we could just use casts to promote smaller values to larger, but what I have doesn't work.  Maybe casting as:
```
(uint64_t) ((int64_t) x)
```
would sign extend properly?

Reply:
Unfortunately it appears that won't do the trick, probably since there's still enough bits in an int64_t for the overflow value to fit.

Reply:
I added those macros to `ofi.h` and updated the check to use them.